crates/noirc_frontend/src/hir/resolution/resolver.rs

```diff
@@ -835,6 +851,108 @@ impl<'a> Resolver<'a> {
+                self.try_eval_array_length_ident(ident.id, span)
+            }
+            ExpressionKind::Prefix(operator) => {
+                let value = self.try_eval_array_length(&operator.rhs)?;
+                match operator.operator {
+                    crate::UnaryOp::Minus => Ok(0 - value),
+                    crate::UnaryOp::Not => Ok(!value),
+                }
+            }
+            ExpressionKind::Infix(operator) => {
+                let lhs = self.try_eval_array_length(&operator.lhs)?;
+                let rhs = self.try_eval_array_length(&operator.rhs)?;
+                match operator.operator.contents {
+                    crate::BinaryOpKind::Add => Ok(lhs + rhs),
+                    crate::BinaryOpKind::Subtract => Ok(lhs - rhs),
+                    crate::BinaryOpKind::Multiply => Ok(lhs * rhs),
+                    crate::BinaryOpKind::Divide => Ok(lhs / rhs),
+                    crate::BinaryOpKind::ShiftRight => Ok(lhs >> rhs),
+                    crate::BinaryOpKind::ShiftLeft => Ok(lhs << rhs),
+                    crate::BinaryOpKind::Modulo => Ok(lhs % rhs),
+                    crate::BinaryOpKind::And => Ok(lhs & rhs),
+                    crate::BinaryOpKind::Or => Ok(lhs | rhs),
+                    crate::BinaryOpKind::Xor => Ok(lhs ^ rhs),
+
+                    crate::BinaryOpKind::Equal
+                    | crate::BinaryOpKind::NotEqual
+                    | crate::BinaryOpKind::Less
+                    | crate::BinaryOpKind::LessEqual
+                    | crate::BinaryOpKind::Greater
+                    | crate::BinaryOpKind::GreaterEqual => {
+                        Err(Some(ResolverError::InvalidArrayLengthExpr { span }))
+                    }
+                }
+            }
+
+            ExpressionKind::Literal(_)
+            | ExpressionKind::Block(_)
+            | ExpressionKind::Index(_)
+            | ExpressionKind::Call(_)
+            | ExpressionKind::MethodCall(_)
+            | ExpressionKind::Constructor(_)
+            | ExpressionKind::MemberAccess(_)
+            | ExpressionKind::Cast(_)
+            | ExpressionKind::For(_)
+            | ExpressionKind::If(_)
+            | ExpressionKind::Tuple(_) => Err(Some(ResolverError::InvalidArrayLengthExpr { span })),
+
+            ExpressionKind::Error => Err(None),
+        }
+    }
+
+    fn try_eval_array_length_ident(
+        &mut self,
+        id: DefinitionId,
+        span: Span,
+    ) -> Result<u128, Option<ResolverError>> {
+        if id == DefinitionId::dummy_id() {
+            return Err(None); // error already reported
+        }
+
+        let definition = self.interner.definition(id);
+        match definition.rhs {
+            Some(rhs) if definition.is_global => self.try_eval_array_length_id(rhs),
+            _ => Err(Some(ResolverError::InvalidArrayLengthExpr { span })),
+        }
+    }
+
+    fn try_eval_array_length_id(&self, rhs: ExprId) -> Result<u128, Option<ResolverError>> {
```

Comment:
I think we could delete `get_fixed_variable_array_length` that is used for resolving the array type and replace its usage with this function